### PR TITLE
Describes how to deal with daemonized processes in dockerfile (Issue #12795)

### PR DESCRIPTION
**Pull request by suchja**
I spent a lot of time figuring this out. Not sure how much of an edge case this is, but would appreciate, if others could benefit from my learnings.

Signed-off-by: Jan Suchotzki jan@suchotzki.de

### CODE REVIEW

File: docs/sources/articles/dockerfile_best-practices.md [hunk 1]
```diff
@@ -55,6 +55,13 @@ container. Decoupling applications into multiple containers makes it much
 easier to scale horizontally and reuse containers. If that service depends on
 another service, make use of [container linking](https://docs.docker.com/userguide/dockerlinks/).
 
+In some cases the process you start might create more processes.
```
Comment (moxiegirl) [inline docs/sources/articles/dockerfile_best-practices.md]: There are some slightly awkward English constructions.  Also, I don't think it is necessary to mention interactive mode as this page is all about the **Dockerfile**.

-->https://gist.github.com/moxiegirl/c5b390b273f9fb732ccc

In some cases, you might start a process that itself creates more processes. If these are daemonized, non-child processes, you should be aware they can
cause unexpected behavior during `docker build`. 

A process you started within a Dockerfile might end before a daemonized
processes it created.  One solution is to check whether the daemonized process
is still running (for example, with `pgrep`) and `sleep` the parent process if it is.

**Comment by moxiegirl** [comment]
ping @docker/maintainers Can someone give this best practice addition a technical eye to make sure its statements are correct?

**Comment by calavera** [comment]
I think this is a rare edge case that should not be documented in the best practices. The build process is designed to be sequential, therefore, using commands that daemonize other processes should not be recommended for building images. You should try to run everything in foreground as much as possible. But this is just my personal opinion.

I might see the value of this information somewhere else, since, as mentioned by @suchja, it's an edge case hard to understand.

**Comment by SvenDowideit** [comment]
I'm not sure i agree with @calavera - this is the place we're telling the user to use only one process - and expanding on that to confirm that we don't mean the service should have no children is good

and then expanding a little further on the gotcha seems worth it - as we have the context, and their attention.

**Comment by jamtur01** [comment]
I think I agree with @calavera primarily but I do think we need to say something - although perhaps not this. Perhaps more a warning/note?

**Comment by icecrime** [comment]
I'm going with @calavera here: it's not a best practice. I'm also gonna go with @jamtur01: it needs a note.

Thank you @suchja for this contribution, however I don't think we should merge as it is. I'm going to close it, but please feel free to submit a new PR which adds a note explaining that it's **not recommended** to (directly or indirectly) invoke daemonized processes in Dockerfile instructions for the reasons you encountered.

**Pull request closed without merging**